pkg/instancetype/instancetype.go

```diff
@@ -357,6 +365,67 @@ func getInstancetypeAPISpec(obj runtime.Object) (interface{}, error) {
+		return k8sfield.NewPath("spec", "instancetype"), fmt.Errorf(insufficientInstanceTypeCPUResourcesErrorFmt, instancetypeSpec.CPU.Guest, preferenceSpec.Requirements.CPU.Guest)
+	}
+	if instancetypeSpec == nil && vmiSpec.Domain.CPU != nil {
+		// FIXME - Move PreferredCPUTopology to a string pointer and default in mutation webhook before removing this duplicate defaulting
+		preferredCPUTopology := instancetypev1alpha2.PreferSockets
+		if preferenceSpec.CPU != nil && preferenceSpec.CPU.PreferredCPUTopology != "" {
+			preferredCPUTopology = preferenceSpec.CPU.PreferredCPUTopology
+		}
+		switch preferredCPUTopology {
+		case instancetypev1alpha2.PreferThreads:
+			if vmiSpec.Domain.CPU.Threads < preferenceSpec.Requirements.CPU.Guest {
+				return k8sfield.NewPath("spec", "template", "spec", "domain", "cpu", "threads"), fmt.Errorf(insufficientVMCPUResourcesErrorFmt, vmiSpec.Domain.CPU.Threads, preferenceSpec.Requirements.CPU.Guest, "threads")
+			}
+		case instancetypev1alpha2.PreferCores:
+			if vmiSpec.Domain.CPU.Cores < preferenceSpec.Requirements.CPU.Guest {
+				return k8sfield.NewPath("spec", "template", "spec", "domain", "cpu", "cores"), fmt.Errorf(insufficientVMCPUResourcesErrorFmt, vmiSpec.Domain.CPU.Cores, preferenceSpec.Requirements.CPU.Guest, "cores")
+			}
+		case instancetypev1alpha2.PreferSockets:
+			if vmiSpec.Domain.CPU.Sockets < preferenceSpec.Requirements.CPU.Guest {
+				return k8sfield.NewPath("spec", "template", "spec", "domain", "cpu", "sockets"), fmt.Errorf(insufficientVMCPUResourcesErrorFmt, vmiSpec.Domain.CPU.Sockets, preferenceSpec.Requirements.CPU.Guest, "sockets")
+			}
+		}
+	}
+	return nil, nil
+}
+
+func checkMemoryPreferenceRequirements(instancetypeSpec *instancetypev1alpha2.VirtualMachineInstancetypeSpec, preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) (*k8sfield.Path, error) {
+	if instancetypeSpec != nil && instancetypeSpec.Memory.Guest.Cmp(preferenceSpec.Requirements.Memory.Guest) < 0 {
+		return k8sfield.NewPath("spec", "instancetype"), fmt.Errorf(insufficientInstanceTypeMemoryResourcesErrorFmt, instancetypeSpec.Memory.Guest.String(), preferenceSpec.Requirements.Memory.Guest.String())
+	}
+
+	if instancetypeSpec == nil && vmiSpec.Domain.Memory != nil && vmiSpec.Domain.Memory.Guest.Cmp(preferenceSpec.Requirements.Memory.Guest) < 0 {
+		return k8sfield.NewPath("spec", "template", "spec", "domain", "memory"), fmt.Errorf(insufficientVMMemoryResourcesErrorFmt, vmiSpec.Domain.Memory.Guest.String(), preferenceSpec.Requirements.Memory.Guest.String())
+	}
+	return nil, nil
+}
+
+func (m *InstancetypeMethods) CheckPreferenceRequirements(instancetypeSpec *instancetypev1alpha2.VirtualMachineInstancetypeSpec, preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) (*k8sfield.Path, error) {
+	if preferenceSpec == nil || preferenceSpec.Requirements == nil {
```

Comment:
Even better, move `instancetypeSpec == nil` here?

Reply:
Both branches are used later so I can't do anything with it here.

Reply:
Isn't this the only place where `checkCPUPreferenceRequirements` and `checkMemoryPreferenceRequirements` are called? You do the `nil` check of `preferenceSpec` here, so why not for `instancetypeSpec` too?

Reply:
Oops, asking stupid questions again today. Totally mistook that `==` for a second `!=`. Nevermind.